clients/android/app/src/main/java/app/lockbook/screen/ShareReceiverActivity.kt

```diff
@@ -0,0 +1,198 @@
+            if (f is CreateFileDialogFragment) {
+                model.refreshOverFolder()
+            }
+        }
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        _binding = ActivityShareReceiverBinding.inflate(layoutInflater)
+        setUpView()
+        setContentView(binding.root)
+
+        when (intent?.action) {
+            Intent.ACTION_SEND_MULTIPLE -> {
+                val receivedUris = intent.getParcelableArrayListExtra<Uri>(Intent.EXTRA_STREAM)
+                if (receivedUris != null) {
+                    uris = receivedUris
+                }
+            }
+            Intent.ACTION_SEND -> {
+                val receivedUri = intent.getParcelableExtra<Uri>(Intent.EXTRA_STREAM)
+                if (receivedUri != null) {
+                    uris.add(receivedUri)
+                }
+            }
+        }
+
+        binding.importButton.setOnClickListener {
+            binding.importProgress.visibility = View.VISIBLE
+            binding.importButton.isEnabled = false
+            lifecycleScope.launch {
+                val fileId = importFromUris()
+                importedFileId.value = fileId
+            }
+        }
+
+        val sharedFilesCount = uris.count()
+        val subTitle = if (sharedFilesCount > 1) {
+            "Importing $sharedFilesCount files"
+        } else if (sharedFilesCount == 1) {
+            "Importing " + getUriFileName(uris[0])
+        } else {
+            ""
+        }
+        binding.toolbar.setSubtitle(subTitle)
+
+        supportFragmentManager.registerFragmentLifecycleCallbacks(
+            fragmentFinishedCallback,
+            false
+        )
+
+        binding.toolbar.setOnMenuItemClickListener {
+            activityModel.launchTransientScreen(TransientScreen.Create(model.currentParent.id))
+            CreateFileDialogFragment().show(
+                supportFragmentManager,
+                CreateFileDialogFragment.TAG
+            )
+            true
+        }
+
+        importedFileId.observe(this) { id ->
+            binding.importProgress.visibility = View.INVISIBLE
+            binding.importButton.isEnabled = true
+
+            if (id !== null) {
+                startActivity(
+                    Intent(this, MainScreenActivity::class.java).apply {
+                        putExtra(IMPORTED_FILE_KEY, id)
+                    }
+                )
+            }
+        }
+    }
+
+    private fun getUriFileName(uri: Uri): String? {
+        var fileName = "untitled"
+        contentResolver.query(uri, null, null, null, null)?.use { cursor ->
+            if (cursor.moveToFirst()) {
+                val displayNameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
+
+                if (displayNameIndex != -1) {
+                    fileName = cursor.getString(displayNameIndex)
+                }
+            }
+        }
+        return fileName
+    }
```

Comment:
Surprised there isn't a simpler way to get an absolute file path, can we check again (I believe I do something like this in other places, exporting a file)

Reply:
Oh, makes sense